Improve date/datetime format handling and customisability

admdly opened the pull request:
Improves date/datetime handling by replacing custom `bb_date` and `bb_datetime` Twig filters with `format_date` / `format_datetime` to better support i18n and customisability.

This adds the `locale`, `timezone`, `date_format`, `time_format` and `datetime_pattern` options under `i18n` to config.php. 

Also resolves #938 by adding a reset button to the admin date picker, and fixing an issue with the Order module not accepting blank dates (thanks to @BelleNottelling for assistance with this).

Note: The datepicker does not currently honour specified formats, and additionally does not support selecting times at all. Fixing it beyond the scope of this PR, but the issue should be noted to be addressed at a later date.

BelleNottelling commented:
Labeled as a breaking change, since this will remove twig functions custom modules may be using

admdly commented:
Update - added patch to `foss-update.php` to replace `bb_date` and `bb_datetime` with the new filters to handle email templates as suggested by @BelleNottelling. 

Also manually reviewed and updated around half of the templates to ensure date formats make sense in context, and to find any cases where the `date` filter was being used. The remaining templates will be reviewed and updated tomorrow. 🙂

BelleNottelling commented on src/library/Box/Tools.php:
```diff
@@ -406,8 +406,11 @@ public function updateConfig()
-        $newConfig['locale_date_format'] = ($currentConfig['locale_date_format'] === '%A, %d %B %G') ? 'l, d F o' : $currentConfig['locale_date_format'];
-        $newConfig['locale_time_format'] = ($currentConfig['locale_time_format'] === ' %T') ? ' G:i:s' : $currentConfig['locale_time_format'];
+
+        $newConfig['i18n']['locale'] ??= 'en_US';
```
We should be bringing over the old values for these when they exist

admdly replied:
Thanks for pointing this out - updated in latest commit. 🙂

admdly commented:
This PR is now ready for review/testing - I will continue testing myself and push any fixes if needed in the meantime, though not major is expected.

BelleNottelling commented on src/di.php:
```diff
@@ -744,7 +753,7 @@
 
     $cookieBBlang = $di['cookie']->get('BBLANG');
-    $locale = !empty($cookieBBlang) ? $cookieBBlang : $di['config']['locale'];
+    $locale = !empty($cookieBBlang) ? $cookieBBlang : $di['config']['i18n']['locale'];
```
Needs a fallback if it's undefined. This will do:
`$locale = !empty($cookieBBlang) ? $cookieBBlang : ($di['config']['i18n']['locale'] ?? 'en_US');`

admdly replied:
Added 🙂 .

BelleNottelling commented on src/di.php:
```diff
@@ -306,6 +307,14 @@
+    // missing required settings.
+    $locale = $config['i18n']['locale'] ?? 'en_US';
+    $timezone = $config['i18n']['timezone'] ?? 'UTC';
+    $date_format = $config['i18n']['date_format'] ? strtoupper($config['i18n']['date_format']) : 'MEDIUM';
```
This won't work, you're trying to reference an undefined var. Will have to wrap them in an isset

``` PHP
$date_format =  isset($config['i18n']['date_format']) ? strtoupper($config['i18n']['date_format']) : 'MEDIUM';
$time_format =  isset($config['i18n']['time_format']) ? strtoupper($config['i18n']['time_format']) : 'SHORT';
```

admdly replied:
I've added an `!empty()` check on those variables as discussed. 🙂

BelleNottelling reviewed (changes requested):
LGTM once my two comments are resolved

BelleNottelling reviewed (approved):
Looks great, I'm glad to move away from custom filters and instead use the twig ones
